nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/bulk/IndexRequestComputation.java

```diff
@@ -0,0 +1,162 @@
+    protected static final long MAX_RECORD_SIZE = 990_000;
+
+    protected static final String INDEX_OPTION = "indexName";
+
+    protected final boolean continueOnFailure;
+
+    protected boolean abort;
+
+    protected BulkRequest bulkRequest;
+
+    protected List<BulkRequest> bulkRequests = new ArrayList<>();
+
+    protected String bucketKey;
+
+    public IndexRequestComputation(boolean continueOnFailure) {
+        super(ACTION_NAME);
+        this.continueOnFailure = continueOnFailure;
+    }
+
+    @Override
+    public void startBucket(String bucketKey) {
+        this.bucketKey = bucketKey;
+        bulkRequests.clear();
+        bulkRequest = new BulkRequest();
+    }
+
+    @Override
+    protected void compute(CoreSession session, List<String> documentIds, Map<String, Serializable> properties) {
+        long now = Timestamp.currentTimeMicros();
+        String indexName = getIndexName(session, properties);
+        DocumentModelList docs = loadDocuments(session, documentIds);
+        ElasticSearchIndexing esi = Framework.getService(ElasticSearchIndexing.class);
+        for (DocumentModel doc : docs) {
+            try {
+                append(new IndexRequest(indexName, DOC_TYPE, doc.getId()).source(esi.source(doc), XContentType.JSON)
+                                                                         .versionType(VersionType.EXTERNAL)
+                                                                         .version(now));
+            } catch (IOException e) {
+                log.error("Cannot build source for document: " + doc.getId(), e);
+                if (! continueOnFailure) {
+                    abort = true;
+                }
+            }
+        }
+    }
+
+    protected void append(IndexRequest indexRequest) {
+        if (bulkRequest.estimatedSizeInBytes() + indexRequest.source().length() > MAX_RECORD_SIZE) {
+            if (bulkRequest.numberOfActions() > 0) {
+                // Create multiple elastic bulk requests when we exceed the record size
+                bulkRequests.add(bulkRequest);
+                bulkRequest = new BulkRequest();
+            } else {
+                log.warn(String.format("Indexing request for doc: %s, is too large: %d, max record size: %d",
+                        indexRequest.id(), indexRequest.source().length(), MAX_RECORD_SIZE));
+            }
+        }
+        bulkRequest.add(indexRequest);
+    }
+
+    @Override
+    public void endBucket(ComputationContext context, int bucketSize) {
+        if (abort) {
+            context.askForTermination();
+            log.error("Terminate computation due to previous error");
+            return;
+        }
+        bulkRequests.add(bulkRequest);
+        String commandId = getCurrentCommand().getId();
+        int i = 0;
+        for (BulkRequest request : bulkRequests) {
+            // rand robin key
```

Comment:
rand robin -> round-robin. But is it really though? Round-robin means going back to 0 after some time, and here you just do `i++`

Reply:
the key is always changing so the partition is assigned randomly and evenly, will fix the comment